File: src/main/java/dev/obscuria/tooltips/client/tooltip/element/effect/RimLightEffect.java
package dev.obscuria.tooltips.client.tooltip.element.effect;

import com.github.bsideup.jabel.Desugar;
import com.google.gson.JsonObject;
import dev.obscuria.tooltips.client.TooltipState;
import dev.obscuria.tooltips.client.render.GuiGraphics;
import dev.obscuria.tooltips.client.tooltip.element.QuadPalette;
import dev.obscuria.tooltips.client.tooltip.particle.GraphicUtils;
import net.minecraft.util.JsonUtils;

import java.util.List;

@Desugar
public record RimLightEffect(QuadPalette outerPalette, QuadPalette innerPalette) implements TooltipEffect {

    public static RimLightEffect fromJson(JsonObject json) {
        return new RimLightEffect(
                QuadPalette.fromJson(JsonUtils.getJsonObject(json, "outer_palette")),
                QuadPalette.fromJson(JsonUtils.getJsonObject(json, "inner_palette")));
    }

    @Override
    public boolean canApply(List<TooltipEffect> effects) {
        return effects.stream().noneMatch(it -> it instanceof RimLightEffect);
    }

    @Override
    public void renderBack(TooltipState state, GuiGraphics graphics, int x, int y, int width, int height) {
        final var pX = x - 3f;
        final var pY = y - 3f;
        final var pWidth = width + 6f;
        final var pHeight = height + 6f;

        final var scale = 0.8f + 0.4f * (float) Math.cos(state.timeInSeconds());
        final var offset = Math.min(pWidth, pHeight) * 0.25f * scale;
        
        graphics.pose().pushMatrix();
        GraphicUtils.drawQuad(graphics, pX, pY, pX + offset, pY + offset, pX + pWidth - offset, pY + offset, pX + pWidth, pY,
                outerPalette.topLeft(), innerPalette.topLeft(), innerPalette.topRight(), outerPalette.topRight());
        GraphicUtils.drawQuad(graphics, pX + offset, pY + pHeight - offset, pX, pY + pHeight, pX + pWidth, pY + pHeight, pX + pWidth - offset, pY + pHeight - offset,
                innerPalette.bottomLeft(), outerPalette.bottomLeft(), outerPalette.bottomRight(), innerPalette.bottomRight());
        GraphicUtils.drawQuad(graphics, pX, pY, pX, pY + pHeight, pX + offset, pY + pHeight - offset, pX + offset, pY + offset,
                outerPalette.topLeft(), outerPalette.bottomLeft(), innerPalette.bottomLeft(), innerPalette.topLeft());
        GraphicUtils.drawQuad(graphics, pX + pWidth - offset, pY + offset, pX + pWidth - offset, pY + pHeight - offset, pX + pWidth, pY + pHeight, pX + pWidth, pY,
                innerPalette.topRight(), innerPalette.bottomRight(), outerPalette.bottomRight(), outerPalette.topRight());
        graphics.pose().popMatrix();
    }
}

File: common/src/main/kotlin/dev/obscuria/tooltips/client/element/effect/RimLightEffect.kt
package dev.obscuria.tooltips.client.element.effect

import com.mojang.serialization.Codec
import com.mojang.serialization.codecs.RecordCodecBuilder
import dev.obscuria.tooltips.client.element.QuadPalette
import dev.obscuria.tooltips.client.renderer.TooltipContext
import dev.obscuria.tooltips.extension.GraphicsExtensions.color
import dev.obscuria.tooltips.extension.GraphicsExtensions.vertex
import net.minecraft.client.gui.GuiGraphics
import net.minecraft.client.renderer.RenderType
import kotlin.math.cos
import kotlin.math.min

class RimLightEffect(
    val outerPalette: QuadPalette,
    val innerPalette: QuadPalette
) : ITooltipEffect {

    override fun codec(): Codec<RimLightEffect> = CODEC

    override fun canApply(effect: List<ITooltipEffect>): Boolean {
        return effect.none { it is RimLightEffect }
    }

    override fun renderBack(graphics: GuiGraphics, context: TooltipContext, x: Int, y: Int, width: Int, height: Int) {

        val pX = x - 3
        val pY = y - 3
        val pWidth = width + 6
        val pHeight = height + 6

        val matrix = graphics.pose().last().pose()
        val buffer = graphics.bufferSource().getBuffer(RenderType.guiOverlay())
        val scale = 0.8 + 0.4 * cos(context.timeInSeconds())
        val offset = min(pWidth, pHeight) * 0.25 * scale

        buffer.vertex(matrix, pX, pY).color(outerPalette.topLeft).endVertex()
        buffer.vertex(matrix, pX + offset, pY + offset).color(innerPalette.topLeft).endVertex()
        buffer.vertex(matrix, pX + pWidth - offset, pY + offset).color(innerPalette.topRight).endVertex()
        buffer.vertex(matrix, pX + pWidth, pY).color(outerPalette.topRight).endVertex()

        buffer.vertex(matrix, pX + offset, pY + pHeight - offset).color(innerPalette.bottomLeft).endVertex()
        buffer.vertex(matrix, pX, pY + pHeight).color(outerPalette.bottomLeft).endVertex()
        buffer.vertex(matrix, pX + pWidth, pY + pHeight).color(outerPalette.bottomRight).endVertex()
        buffer.vertex(matrix, pX + pWidth - offset, pY + pHeight - offset).color(innerPalette.bottomRight).endVertex()

        buffer.vertex(matrix, pX, pY).color(outerPalette.topLeft).endVertex()
        buffer.vertex(matrix, pX, pY + pHeight).color(outerPalette.bottomLeft).endVertex()
        buffer.vertex(matrix, pX + offset, pY + pHeight - offset).color(innerPalette.bottomLeft).endVertex()
        buffer.vertex(matrix, pX + offset, pY + offset).color(innerPalette.topLeft).endVertex()

        buffer.vertex(matrix, pX + pWidth - offset, pY + offset).color(innerPalette.topRight).endVertex()
        buffer.vertex(matrix, pX + pWidth - offset, pY + pHeight - offset).color(innerPalette.bottomRight).endVertex()
        buffer.vertex(matrix, pX + pWidth, pY + pHeight).color(outerPalette.bottomRight).endVertex()
        buffer.vertex(matrix, pX + pWidth, pY).color(outerPalette.topRight).endVertex()
    }

    companion object {

        val CODEC: Codec<RimLightEffect> = RecordCodecBuilder.create { codec ->
            codec.group(
                QuadPalette.CODEC
                    .fieldOf("outer_palette")
                    .forGetter(RimLightEffect::outerPalette),
                QuadPalette.CODEC
                    .fieldOf("inner_palette")
                    .forGetter(RimLightEffect::innerPalette)
            ).apply(codec, ::RimLightEffect)
        }
    }
}